handlers/user.go

```diff
@@ -1298,109 +1336,197 @@ func UserUpdateEmailPost(w http.ResponseWriter, r *http.Request) {
-		session.AddFlash("Error: Email already exists please choose a unique email")
+	}
+
+	if emailExists {
+		session.AddFlash("Error: Email already exists, please choose a unique email")
 		session.Save(r, w)
 		http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
 		return
 	}
 
-	var rateLimitError *types.RateLimitError
-	err = sendEmailUpdateConfirmation(user.UserID, email)
+	// everything is fine, send confirmation email
+
+	err = sendEmailUpdateConfirmation(user.UserID, newEmail)
 	if err != nil {
-		logger.Errorf("error sending confirmation-email: %v", err)
-		if errors.As(err, &rateLimitError) {
-			session.AddFlash(fmt.Sprintf("Error: The ratelimit for sending emails has been exceeded, please try again in %v.", err.(*types.RateLimitError).TimeLeft.Round(time.Second)))
-		} else {
-			session.AddFlash(authInternalServerErrorFlashMsg)
-		}
+		utils.LogError(err, "error sending email-change confirmation email", 0, map[string]interface{}{"userID": user.UserID, "newEmail": newEmail})
+		session.AddFlash(authInternalServerErrorFlashMsg)
 		session.Save(r, w)
 		http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
 		return
 	}
 
-	session.AddFlash("Verification link sent to your new email " + email)
+	session.AddFlash("Verification link sent to your new email " + newEmail)
 	session.Save(r, w)
 	http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
 }
 
+func sendEmailUpdateConfirmation(userId uint64, newEmail string) error {
+	emailConfirmationHash := utils.RandomString(40)
+
+	_, err := db.FrontendWriterDB.Exec("UPDATE users SET email_confirmation_hash = $1, email_change_to_value = $2, email_confirmation_ts = TO_TIMESTAMP($3) WHERE id = $4", emailConfirmationHash, newEmail, time.Now().Unix(), userId)
+	if err != nil {
+		return fmt.Errorf("error updating db data for user %v for email change: %w", userId, err)
+	}
+
+	subject := fmt.Sprintf("%s: Verify your email-address", utils.Config.Frontend.SiteDomain)
+	msg := fmt.Sprintf(`To update your email on %[1]s please verify it by clicking this link:
+
+https://%[1]s/settings/email/%[2]s
+
+This link will expire in 30 minutes.
+
+Best regards,
+
+%[1]s
+`, utils.Config.Frontend.SiteDomain, emailConfirmationHash)
+	err = mail.SendTextMail(newEmail, subject, msg, []types.EmailAttachment{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
 // ConfirmUpdateEmail confirms and updates the email address of the user. Given an update link the email in the db is changed.
 func UserConfirmUpdateEmail(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	hash := vars["hash"]
 
-	_, session, err := getUserSession(r)
+	sessionUser, session, err := getUserSession(r)
 	if err != nil {
-		logger.Errorf("error retrieving session: %v", err)
+		utils.LogError(err, "error retrieving session for email update confirmation", 0, map[string]interface{}{"hash": hash})
 		http.Error(w, "Internal server error", http.StatusInternalServerError)
 		return
 	}
+	if !sessionUser.Authenticated {
```

Comment:
I was a bit on the fence on what to do here, since at the end of that function all user sessions (for that user) get purged and the user gets logged out. I was mainly concerned with what happens if the user clicks on the confirmation link, but is logged into a different account on that device.
With my new changes, not being authenticated or being logged into a different user account will not prevent an email change.